File: code-ninja-course/JS/src/solid/StudentService.js
class StudentService {

	addIfValid(emailAddress, universityId){
		//SRP
		Console.log("Log: Start add student with email " + emailAddress);

		if (emailAddress === "") 	{
			return false;
		}

		var studentRepository = new StudentRepository();

		if (studentRepository.exists(emailAddress))	{
			return false;
		}

		var universityRepository = new UniversityRepository();

		var university = universityRepository.getByID(universityId);

		var student = new Student(emailAddress, universityId);

		//SRP, OCP
		type = university.getPackageType();
		if ( type === PackageTypeEnum.Standard) {
			student.setMonthlyEbookAllowance(10);
		}
		else if (type === PackageTypeEnum.Premium) {
			student.setMonthlyEbookAllowance (10 * 2);
		}

		studentRepository.add(student);

		//SRP
		console.log("Log: End add student with email " + emailAddress);

		return true;//CQS

	}
	
	getStudentsByUniversity() {
		throw "NotImplemeneted";
	}
	
	getStudentsByCurrentlyBorrowedEbooks() {
		throw "NotImplemented";
	}
}

File: code-ninja-course/C/src/solid/StudentService.c
/*
 * StudentService.c
 *
 *  Created on: 17 ????? 2018
 *      Author: Gil
 */
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>

#include "PackageType.h"
#include "University.h"
#include "Student.h"
#include "UniversityRepository.h"
#include "StudentRepository.h"

void addStudentToRepository(struct StudentRepository* repository, struct Student* student);
struct University* getUniversityByID(struct UniversityRepository* repository, int universityID);
struct Student* createStudent (const char* _emailAddress, int _universityID);
void setStudentMonthlybookAllowance(struct Student* student, int allowance);

int addStudent(char* emailAddress, int universityID)
{
	//SRP
	printf("Log: Start add student with email %s", emailAddress);

	if (emailAddress == NULL) 	{
		return -1;
	}

	struct StudentRepository studentRepository;

	struct UniversityRepository universityRepository;

	struct University* university =
			getUniversityByID(&universityRepository,
					universityID);


	struct Student* student =
			createStudent(emailAddress, universityID);

	//SRP, OCP
	PackageType type = university->packageType;
	if ( type == Standard) {
		setStudentMonthlybookAllowance(student,10);
	}
	else
		if (type == Premium) {
		setStudentMonthlybookAllowance(student,2*10);
	}

	addStudentToRepository(
			&studentRepository,
			student);

	//SRP
	printf("Log: End add student with email %s",emailAddress);

	return 0;//CQS

}

struct Student* getStudentsByUniversity(int universityID) {
	return NULL;
}

struct Student* getStudentsByCurrentlyBorrowedbooks() {
	return NULL;
}


